src/shared/watchdog.c

```diff
@@ -462,20 +496,26 @@ static usec_t watchdog_calc_timeout(void) {
                 return watchdog_timeout;
 }
 
-usec_t watchdog_runtime_wait(void) {
+usec_t watchdog_runtime_wait(unsigned divisor) {
         usec_t timeout = watchdog_calc_timeout();
         if (!timestamp_is_set(timeout))
                 return USEC_INFINITY;
 
-        /* Sleep half the watchdog timeout since the last successful ping at most */
-        if (timestamp_is_set(watchdog_last_ping)) {
+        /* Sleep watchdog timeout / divisor since the last successful ping attempt at most,
+         * or the fifth of that if the ping failed. */
+
+        usec_t ts = MAX(timestamp_is_set(watchdog_last_good_ping) ? watchdog_last_good_ping : 0u,
+                        timestamp_is_set(watchdog_last_bad_ping) ? watchdog_last_bad_ping : 0u);
+        if (ts > 0) {
                 usec_t ntime = now(CLOCK_BOOTTIME);
+                if (watchdog_last_bad_ping == ts)
```

Comment:
not loving this kind of test that works from the results backwards...

Reply:
Yeah. But the alternative would be to either repeat the logic or have a lot more verbose code… I reversed the order in the check, I hope that that makes it more readable.